x-pack/filebeat/input/s3/input.go

```diff
@@ -531,6 +531,8 @@ func (c *s3Context) Fail(err error) {
 }
 
 func (c *s3Context) done() {
```

Comment:
`Fail()` calls `done()` and holds the lock. That will result in a deadlock. I didn't look at the rest of the file, but wanted to call this out now. I can probably look at the rest of it tomorrow if any assistance is needed.

Reply:
I agree. That needs to be resolved.

Considering that we just want the latest error for now, we should be fine with the following change as only the counter needs to be protected against race conditions:
```suggestion
func (c *s3Context) Fail(err error) {
	// only care about the last error for now
	// TODO: add "Typed" error to error for context
	c.err = err
	c.done()
}
```

Reply:
if we have a race in 'done' due to concurrent calls messing with the reference counter, I wonder if Fail can be triggered concurrently as well. Maybe it would be better to have Failed like this (we do run tests with race detector enabled):

```
func (c *s3Context) Fail(err error) {
	// only care about the last error for now
	// TODO: add "Typed" error to error for context
	c.mux.Lock()
	c.err = err
	c.mux.Unlock()
	c.done()
}
```

or:

```
func (c *s3Context) Fail(err error) {
	c.setError(err)
	c.done()
}

func (c *s3Context) SetError(err error) {
	// only care about the last error for now
	// TODO: add "Typed" error to error for context
	c.mux.Lock()
	c.err = err
	c.mux.Unlock()
}
```

Reply:
I have been running with the fix provided in the latest commit for about 4 days now. The ingest pipeline is processing about 1.5 million log lines every 5 minutes and so far it has been stable. Without the fix, the software kept crashing after processing a few files from S3.